File: Fast & Slow Pointer/ReorderList.java
public class ReorderList {
    public static void main(String[] args) {
        ReorderList solution = new ReorderList();
        ListNode one = new ListNode(1);
        ListNode two = new ListNode(2);
        ListNode three = new ListNode(3);
        ListNode four = new ListNode(4);
        one.next = two;
        two.next = three;
        three.next = four;
        ListNode ans = solution.input(one);
        System.out.println(ans);
    }

    public ListNode input(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode curr = slow.next;
        slow.next = null;
        ListNode prev = null;
        ListNode forw = null;
        while (curr != null) {
            forw = curr.next;
            curr.next = prev;
            prev = curr;
            curr = forw;
        }
        while (prev != null) {
            ListNode p = head.next;
            ListNode q = prev.next;
            head.next = prev;
            prev.next = p;
            prev = q;
            head = p;
        }

        // ListNode slow = head;
        // ListNode fast = head;
        // while (fast != null && fast.next != null) {
        // slow = slow.next;
        // fast = fast.next.next;
        // }
        // ListNode prev = null;
        // while (slow != null) {
        // ListNode temp = slow.next;
        // slow.next = prev;
        // prev = slow;
        // slow = temp;
        // }
        // ListNode p1 = head;
        // ListNode p2 = prev;
        // while (p1 != null && p2 != null) {
        // ListNode next1 = p1.next;
        // ListNode next2 = p2.next;
        // p1.next = p2;
        // p2.next = next1;
        // p1 = next1;
        // p2 = next2;
        // }
        return head;
    }
}

File: Fast & Slow Pointer/ReorderList.js
class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}
var ReorderList = function(head) {
    if(!head||!head.next) return;
    const nodearr=[];
    let cur=head;
    while(cur){
        nodearr.push(cur);
        cur=cur.next;
        nodearr[nodearr.length-1].next=null;
    }
    let begin=0,end=nodearr.length-1;
    cur=null;
    while(begin<end){
        if(cur){
            cur.next=nodearr[begin];
        }
        nodearr[begin].next=nodearr[end];
        cur=nodearr[end];
        begin++;
        end--;
        if(begin==end){
            cur.next=nodearr[begin];
        }
    }
    return head;
};
const head = new ListNode(1);
const second = new ListNode(2);
const third = new ListNode(3);
const fourth = new ListNode(4);

head.next = second;
second.next = third;
third.next = fourth;
fourth.next = null;

const result = ReorderList(head);
console.log(result);
